PayItForward.Data/Repositories/EfGenericRepository.cs

```diff
@@ -0,0 +1,144 @@
+﻿namespace PayItForward.Data
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using Microsoft.EntityFrameworkCore;
+    using PayItForward.Data.Models;
+    using PayItForward.Data.Models.Abstraction;
+
+    public class EfGenericRepository<T, TKey> : IRepository<T, TKey>
+        where T : BaseModel<TKey>
+    {
+        public EfGenericRepository(PayItForwardDbContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentException("An instance of DbContext is required to use this repository.", "context");
+            }
+
+            this.Context = context;
+            this.DbSet = this.Context.Set<T>();
+        }
+
+        protected DbSet<T> DbSet { get; set; }
+
+        protected PayItForwardDbContext Context { get; set; }
+
+        public async Task<T> GetByIdAsync(TKey id)
+        {
+            return await this.DbSet.FindAsync(id);
+        }
+
+        public virtual T GetById(TKey id)
+        {
+            return this.DbSet.Find(id);
+        }
+
+        public void Add(T entity)
+        {
+            this.ChangeEntityState(entity, EntityState.Added);
+        }
+
+        public void Update(T entity)
+        {
+            var entry = this.Context.Entry(entity);
+            if (entry.State == EntityState.Detached)
+            {
+                this.DbSet.Attach(entity);
+            }
+
+            entry.State = EntityState.Modified;
+        }
+
+        public IQueryable<T> GetAll()
+        {
+            return this.DbSet;
+        }
+
+        public Task<IEnumerable<T>> GetAllAsync()
+        {
+            return Task.FromResult(this.DbSet.AsEnumerable());
+        }
+
+        // to be tested
+        public async Task UpdateAsync(T entity)
```

Comment:
Ypou dont need this method, as you have update that changes the state and after that you can call async save changes to make this saving async